exercise/models.py

```diff
@@ -129,24 +129,28 @@ def __str__(self):
     emails = models.ManyToManyField(ExerciseEmail, blank=True)
 
     def set_email_reveal_times(self):
-        if not self.emails.all():
+        emails = self.emails.all()
+        if not emails:
             return
+
         # Set some emails based on a threshold to zero time initially.
-        threshold_emails = int(ceil((self.emails.count() * settings.REVEAL_TIME_ZERO_THRESHOLD)))
+        received_emails_count = int(ceil((len(emails) * settings.REVEAL_TIME_ZERO_THRESHOLD)))
 
-        while threshold_emails > 0:
-            email = self.emails.all()[randint(0, self.emails.count()-1)]
-            ExerciseEmailRevealTime.objects.get(exercise_id=self.id, email_id=email.id).set_reveal_time(0)
-            threshold_emails -= 1
+        while received_emails_count > 0:
+            email = emails[randint(0, len(emails)-1)]
+            email_properties = ExerciseEmailProperties.objects.filter(exercise_id=self.id, email_id=email.id).first()
+            if email_properties:
+                email_properties.set_reveal_time(0)
+            received_emails_count -= 1
 
         # Remaining emails without a set time to be set to a random time
-        for email in self.emails.all():
+        for email in emails:
```

Comment:
without `.pop()` we still have that risk to pick the same email in the `while` loop
Querysets can be converted to lists with `list(queryset)`

Reply:
I'll follow up on this one after merging

Reply:
Ah yes but it won't do anything if its already set to zero. It's not a huge deal I guess... just runs through the loop ignoring anything already set to zero or any value for that matter.

Reply:
Yeah it's the "accessing the element" in the while loop that has a chance to pick the same, then you end up with less emails with 0 reveal time than expected.